src/particles.cpp

```diff
@@ -215,6 +216,8 @@ void ServerParticleTexture::deSerialize(std::istream &is, u16 protocol_ver,
 
 	animated = !!(flags & FlagT(ParticleTextureFlags::animated));
 	blendmode = BlendMode((flags & FlagT(ParticleTextureFlags::blend)) >> 1);
+	if (blendmode >= BlendMode::BlendMode_END)
+		throw SerializationError("invalid blend mode");
```

Comment:
Is is better to error and not show the particle instead of defaulting to - say - `clip`?
Reason for this question is that mods would have to check each client protocol version to ensure that the particle is shown.

Reply:
> Is is better to error [...]

Yes. This check only fails if the server is sending us a blend mode we don't understand. Because the server knows our protocol version and the server code is written correctly, this should never happen. If this check is triggered, the server is broken or malicious.

> Reason for this question is that mods would have to check each client protocol version to ensure that the particle is shown.

They don't have to, it's the Luanti server's business to convert new blend modes to old blend modes (for example "clip" is converted to "alpha", see above) for older clients that don't understand them (or to drop the particles altogether if there's no chance that it'll look remotely correct).

---

If it bothers you, I can get rid of this check or move it to a separate commit; it is ultimately unnecessary for the "clip" feature. [^1] (Though the absence of this check bothers me. We're inviting invalid enum values here, which may then cause undefined behavior e.g. if values are not initialized by switches. Network deserialization code should always validate so that invariants aren't violated further down the line.)

[^1]: I believe I added it because I had originally touched the `switch` more.

Reply:
server-side deals with this, it's the code change above this one

Reply:
Thanks for pointing out the server code above. I've seen the code above, but apparently didn't read the comment that would explain it.
Comment resolved. No action needed.